db/wal_edit.cc

```diff
@@ -172,4 +172,41 @@ Status WalSet::DeleteWals(const WalDeletions& wals) {
 
 void WalSet::Reset() { wals_.clear(); }
 
+Status WalSet::CheckWals(
+    Env* env,
+    const std::unordered_map<WalNumber, std::string>& logs_on_disk) const {
+  assert(env != nullptr);
+
+  Status s;
+  for (const auto& wal : wals_) {
+    const uint64_t log_number = wal.first;
+    const WalMetadata& wal_meta = wal.second;
+
+    if (logs_on_disk.find(log_number) == logs_on_disk.end()) {
+      std::stringstream ss;
+      ss << "Missing WAL with log number: " << log_number << ".";
+      s = Status::Corruption(ss.str());
+      break;
+    }
+
+    if (wal_meta.HasSize()) {
+      uint64_t log_file_size = 0;
+      s = env->GetFileSize(logs_on_disk.at(log_number), &log_file_size);
+      if (!s.ok()) {
+        break;
+      }
+      if (wal_meta.GetSizeInBytes() != log_file_size) {
```

Comment:
Per our discussion offline, I think this can be relaxed. The on-disk WAL after a crash can be larger than what was recorded in MANIFEST.
Also, depending on when you write the first `WalEdit` for a WAL file, we should take different actions.
- If you write and sync MANIFEST when you first created a WAL file. Then it is possible for a WAL to be recorded in MANIFEST but not on disk.
- If you write and sync MANIFEST when you sync the WAL, then it's different.

Reply:
Added a "synced" field to WalMetadata, if the WAL or WAL directory is not synced, then we skip checking the WAL.

Reply:
Is the addition of "synced" field pushed?

Reply:
added syncing related information to WalMetadata: https://github.com/facebook/rocksdb/pull/7414